### ISSUE TITLE
Whisper: upstream feature gaps & performance regressions vs mlx-audio Python

### ISSUE THREAD

**Issue by SRjoeee** [findit-studio/mlxrs#317]
## Whisper: upstream feature gaps & performance regressions vs mlx-audio Python

Deep audit of the Whisper STT implementation (`audio/stt/models/whisper/`) compared against the upstream **mlx-audio Python** (`mlx_audio/stt/models/whisper/`). Each issue was verified at the code level by cross-referencing both codebases.

> **Scope**: Whisper model only. wav2vec2/CTC, DSP, generic STT traits, and the streaming framework (Swift-port) are out of scope.

---

### Background

mlxrs is a 1:1 Rust port of mlx-audio. The Whisper port faithfully reproduces the core decode loop (greedy, temperature fallback, timestamp rules, language detection, seek loop), the audio frontend, config/weight validation, and the serialization layer. The **validation layer** (config Extent caps, shape pre-checks, element-count budgets) exceeds what upstream provides.

However, the port has **10 upstream features that were deferred or dropped**, and **2 performance regressions** introduced by the porting approach. Upstream mlx-audio Python already implements all of these.

---

#### 1. Logit filters run on CPU — GPU->CPU round-trip every decode step

**Upstream Python** keeps logits as `mx.array` (GPU) throughout:

```python
# SuppressBlank.apply (decoding.py:356-359) — GPU add
def apply(self, logits, tokens):
    if tokens.shape[1] == self.sample_begin:
        return logits + self.mask    # self.mask is mx.array, stays on GPU
    return logits

# SuppressTokens.apply (decoding.py:368-369) — GPU add
def apply(self, logits, tokens):
    return logits + self.mask

# ApplyTimestampRules.apply (decoding.py:383-442)
mask = np.zeros(...)            # build on CPU
mask = mx.array(mask)           # move to GPU
return logits + mask            # final add on GPU

# GreedyDecoder.update — GPU argmax
next_tokens = logits.argmax(axis=-1)  # mx.argmax on GPU
```

**Rust port** pulls the entire vocab row to CPU every step:

```rust
// decoding.rs:867-885
fn last_position_row(logits3d: &Array) -> Result<Vec<f32>> {
    // ... slice + reshape ...
    row.to_vec::<f32>()    // materializes on GPU, copies to CPU
}

// All LogitFilter::apply operate on &mut [f32] (CPU slices)
trait LogitFilter {
    fn apply(&self, logits: &mut [f32], tokens: &[u32]);  // CPU
}

// argmax_slice — CPU loop
fn argmax_slice(xs: &[f32]) -> u32 {
    for (i, &x) in xs.iter().enumerate() { /* CPU */ }
}
```

**Impact**: ~51K floats (`n_vocab`) transferred GPU->CPU->GPU every decoded token, up to ~224 times per 30-second segment. For `temperature > 0`, the row is additionally reconstructed as an `Array` to call `categorical_sampling` back on GPU. This is the decode hot path.

**Suggested fix**: Port the logit filters to operate on `Array` with `ops::arithmetic::add` / `ops::logical::select` instead of `&mut [f32]` slices. Pre-compute suppress masks as `Array` values (like upstream), then `logits + mask` stays on GPU. Argmax via `ops::reduction::argmax`.

---

#### 2. No quantized model support (upstream has `nn.quantize`)

**Upstream Python** (`whisper.py:659-676`):
```python
quantization = config.pop("quantization", None)
if quantization is not None:
    class_predicate = (
        lambda p, m: isinstance(m, (nn.Linear, nn.Embedding))
        and f"{p}.scales" in weights
    )
    nn.quantize(model, **quantization, class_predicate=class_predicate)
```

**Rust port** (`model.rs:287-316`): `from_weights()` has no `quantization` parameter. Quantized checkpoints (INT8/INT4) either fail to load or load at full precision, using 2-4x more memory.

**Suggested fix**: Consume the crate's existing `PerLayerQuantization` type in `audio/load.rs` during weight loading. Dequantize quantized weights (scales + biases) before building Linear/Embedding layers.

---

### :red_circle: PORT-INDUCED Feature Gaps (upstream has, port deferred)

All 8 features below exist in upstream mlx-audio Python and were deferred during the Rust port.

#### 3. Word-level timestamps (DTW)

**Upstream**: `timing.py` (328 lines) — full DTW alignment, `find_alignment()`, `add_word_timestamps()`, `merge_punctuations()`. Called from `whisper.py:1151` when `word_timestamps=True`.

**Port status**: Explicitly deferred. Attention weights (`qk`/`cross_qk`) are dropped inside `qkv_attention` (`layers.rs:174`), making DTW impossible without re-plumbing.

**Dependency**: Unblocks #4 (hallucination suppression) and #7 (AlignAtt streaming).

#### 4. Hallucination suppression

**Upstream**: `whisper.py:1171-1237` — `word_anomaly_score()`, `is_segment_anomaly()`, `next_words_segment()`. Parameter: `hallucination_silence_threshold`. Uses word-level timing from #3.

**Port status**: Deferred. Requires word timestamps.

#### 5. Best-of-N sampling

**Upstream**: `decoding.py:126` — `best_of: Optional[int]` in `DecodingOptions`. `decoding.py:212-235` — `MaximumLikelihoodRanker` with `length_penalty`. Works with `temperature > 0`.

**Port status**: `best_of` field omitted from `DecodingOptions`. No ranker.

#### 6. Batch decoding

**Upstream**: `decoding.py:634-723` — `DecodingTask.run` handles `n_audio > 1` batch dimension throughout.

**Port status**: `encoder.rs:221-227` enforces `batch == 1`. `decoding.rs:14`: "mlxrs decodes one utterance (n_audio = 1) greedily (n_group = 1)".

**Dependency**: Unblocks #5 (best-of-N needs `n_group > 1`).

#### 7. Cross-attention weight extraction

**Upstream**: `whisper.py:358-359` — `MultiHeadAttention.__call__` returns `(out, kv_cache, qk)`. `decoding.py:177-189` — `logits_with_cross_qk()`. Used by timing.py and streaming.py.

**Port status**: `layers.rs:174` — "The reference's third `qk` return (attention weights, for word-timing DTW) is dropped here; that is a deferred feature."

**Dependency**: Unblocks #3 (DTW), #4 (hallucination), #7 (streaming).

#### 8. AlignAtt streaming decoder

**Upstream**: `streaming.py` (276 lines) — `StreamingDecoder` class, AlignAtt algorithm (arxiv:2211.00895), `get_most_attended_frame()`, `should_emit()`. `whisper.py:1304-1380` — `generate_streaming()` generator.

**Port status**: mlxrs has a generic streaming framework (ported from mlx-audio-swift, NOT the Python AlignAtt). No Whisper-specific streaming implementation.

#### 9. `condition_on_previous_text` (seek-loop prompt threading)

**Upstream**: `whisper.py:801` — `condition_on_previous_text: bool = True`. `whisper.py:1279-1281` — feeds previous segment's tokens as prompt to the next window.

**Port status**: `decoding.rs:100` has `prompt: Vec<u32>`, but the seek loop (`transcribe()`) does NOT feed previous windows' tokens back. Each 30-second window decodes independently.

#### 10. `clip_timestamps` (time-range filtering)

**Upstream**: `whisper.py:807` — `clip_timestamps: Union[str, List[float]] = "0"`. `whisper.py:915-931` — parses into seek_clips, processes only specified time ranges.

**Port status**: Not present. `transcribe()` processes the entire audio.

---

### :yellow_circle: UPSTREAM ISSUE (both Python and Rust have the same limitation)

These are NOT port bugs — upstream mlx-audio Python also lacks them. Listed for completeness.

| Issue | Status |
|-------|--------|
| No fast SDPA (manual `q @ k` -> `softmax` -> `w @ v` in both) | Upstream limitation |
| No KV cache quantization (full-precision `(k, v)` in both) | Upstream limitation |
| No beam search (`NotImplementedError` in upstream) | Upstream limitation |

---

### :white_check_mark: Verified NOT bugs (initial false positives)

| Claim | Status |
|-------|--------|
| `sample_len` off-by-one | Both produce exactly `sample_len` tokens (identical loop structure) |
| Mask offset `k_ctx - n_ctx` underflow | Theoretical only; Rust's offset-aware masking is actually MORE correct than upstream's flat `mask[:n_ctx, :n_ctx]` |
| Mel uses log10 | Both use `log10()` — correct |
| Sinusoids f32 precision | Both compute in f32 then cast to f16 — correct |
| KV cache clone per step | `try_clone()` is O(1) refcount bump, not data copy |

---

### Recommended implementation order

The features form a dependency chain:

```
[7] cross_qk extraction (layers.rs — return qk from qkv_attention)
 └──→ [3] word timestamps / DTW (new timing.rs)
       └──→ [4] hallucination suppression
 └──→ [8] AlignAtt streaming

[6] batch dimension support in decoder
 └──→ [5] best-of-N sampling

[9] seek-loop prompt threading (independent)

[1] logit filters on GPU (independent, highest perf impact)
[2] quantized model loading (independent, memory impact)
```

Suggested phasing:

1. **Phase 1** — `cross_qk` extraction + logit filters to GPU (unblocks DTW + biggest perf win)
2. **Phase 2** — Word timestamps (DTW) + hallucination suppression (quality)
3. **Phase 3** — Batch decoding + best-of-N sampling (accuracy)
4. **Phase 4** — Seek-loop prompt threading + AlignAtt streaming + quantized loading
